### ISSUE TITLE
ASSERT tool.drcachesim.scattergather test: tracer.cpp:394: towrite <= ipc_pipe.get_atomic_write_size() && towrite > 0

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#5329]
tool.drcachesim.scattergather failed:

https://github.com/DynamoRIO/dynamorio/runs/5077131960?check_suite_focus=true
```
ASSERT FAILURE: /home/runner/work/dynamorio/dynamorio/clients/drcachesim/tracer/tracer.cpp:394: towrite <= ipc_pipe.get_atomic_write_size() && towrite > 0 ()
```

I don't remember ever seeing that assert before on any drcachesim test.

**Comment by derekbruening** [comment]
This was the 32-bit x86 test

**Comment by derekbruening** [comment]
Happened again: https://github.com/DynamoRIO/dynamorio/runs/5131569686

**Comment by abhinav92003** [comment]
I ran the 32-bit test 1000 times  on my machine, and couldn't reproduce the failure:

```
$ ctest -VV -R 'tool.drcachesim.scattergather' --repeat-until-fail 1000
...
1/1 Test #284: code_api|tool.drcachesim.scattergather ...   Passed    1.51 sec
 
The following tests passed:
	code_api|tool.drcachesim.scattergather
 
100% tests passed, 0 tests failed out of 1
 
Total Test time (real) = 1547.39 sec
```

I can try running it on the Github Actions runner using `tmate`. In the past, we've had failures which reproduce only on GA.

**Comment by abhinav92003** [comment]
I ran this 1000 times on a GA runner, no failure encountered. Maybe it reproduces on only some runners? If that is so, I'll need to try this multiple times.

```
ctest -VV -R 'tool.drcachesim.scattergather' --repeat-until-fail 1000
...
1/1 Test #286: code_api|tool.drcachesim.scattergather ...   Passed    2.06 sec

The following tests passed:
        code_api|tool.drcachesim.scattergather

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 2024.27 sec
```

**Comment by dolanzhao** [comment]
My PR #5515 failed to pass test 286 multiple times on the x86_32 action.
https://github.com/DynamoRIO/dynamorio/runs/7048743130?check_suite_focus=true

**Comment by abhinav92003** [comment]
> Looks like the following tests are failing deterministically on my machine (and Derek's) now:

The failure I found on my machine was due to a different error, which I fix in #5764. The original assert failure in this issue shows up on our Ubuntu22 CI though.

**Comment by derekbruening** [comment]
This test can fail with this pipe assert 3x in a row, failing even with the retry-3x feature from #2204 for #5873
